### ISSUE TITLE
Emails received in incorrect order

### ISSUE THREAD

**Issue by MikeGuest** [k9mail/k-9#2204]
### Expected behavior
Emails in inbox should be in date order. This is true with other email clients including Gmail and outlook. Behaviour seen intermittently for years but now seems to affect every mail. Love k9 mail but it is no use to me if I can't rely on receiving all my messages.

### Actual behavior
New emails from major Uk ISP virgin media are arriving out of sequence or sometimes not shown at all if there are more than 25 ahead of them of varying ages.

### Steps to reproduce
1. Use k9 mail on virgin media Uk with imap
2.
3.

### Environment
K-9 Mail version: latest as of 7/2/17

Android version: 5.1

Account type (IMAP, POP3, WebDAV/Exchange):
IMAP

Have a screenshot of required. Happy to send

**Comment by philipwhiuk** [comment]
Can you provide the **Date** header from a few of the emails that are shown in the incorrect order.

**Comment by ArchangeGabriel** [comment]
I had a similar issue (but the other way round, message were on top) with this Date header for instance:
```
Date: , 5 Dec 2016 15:17:10 +0200 (CEST)
```
K-9 replaced it with the reception date (i.e. when retrieved by K-9) for sorting.

**Comment by philipwhiuk** [comment]
Yeah - hence why I asked. The problem is we can't support every malformed date header (and we dislike doing work-arounds for other people's bad implementations of email).

**Comment by MikeGuest** [comment]
Top 5 date headers in my inbox are

**Comment by MikeGuest** [comment]
Date: Tue, 7 Feb 2017 16:41:55 +0100 (CET)
Date: Mon, 6 Feb 2017 11:17:44 -0500 (EST)
Date: Tue, 7 Feb 2017 11:13:42 -0500 (EST)
Date: Tue, 7 Feb 2017 07:56:44 +0000
Date: Mon, 6 Feb 2017 12:37:03 +0100 (CET)
Date: Tue, 07 Feb 2017 10:01:11 -0600

Yes I know that's 6 but I wanted to ensure a couple of examples of dates out order. Thanks for taking an interest

**Comment by marmistrz** [comment]
We have the same problem with the following `Date` header.
```
Date: 31 Oct 17 12:13:52
```
This date is replaced by the reception date, in this case 3rd Jan. 

This seems to affect all mails imported by gmail from `tlen.pl`.

**Comment by philipwhiuk** [comment]
Without a zone it's not a valid `Date` header. That's the problem.

I tried a test using MIME4J's lenient parser, but even that doesn't like it. The lenient parser still requires a timezone.

One possibility is to patch MIME4J to add `"dd MMM yy HH:mm:ss` as another, final lenient option, then switch K-9 to use it. Not sure whether they will go for that idea though.

@cketti ?

**Comment by cketti** [comment]
Looks like a duplicate of #1356

**Comment by marmistrz** [comment]
Yeah, that's similar, except that here we have spaces instead of hyphens.

I see your reasoning about _not being able to support every non-standard date format_, still the end-users expect the mail client to _just work_. If they encounter such problems, they'll switch to proprietary solutions like gmail (handles that correctly) or the built-in samsung android client (handles that correctly as well).

Here the situation is in fact more difficult: this mail was downloaded by gmail via their POP3 import from tlen.pl. This means that even if tlen.pl fixed their date format, the old messages downloaded by gmail are already tainted and they'll remain as they are.

It may be worthwhile to see what Thunderbird uses to parse the dates.

**Comment by philipwhiuk** [comment]
If you read more than the opening maxim of the article you find:

> For example, a defective implementation that sends non-conforming messages might be used only with implementations that tolerate those deviations from the specification until, possibly several years later, it is connected with a less tolerant application that rejects its messages. In such a situation, identifying the problem is often difficult, and deploying a solution can be costly. Rose therefore recommended "explicit consistency checks in a protocol ... even if they impose implementation overhead".

This is exactly what has happened here.

**Comment by MikeGuest** [comment]
It's good to have decent standards, however as an end user I've largely moved away from k9 mail into outlook. It's not because I prefer outlook, that has its own problems, it's about being able to reliably access all my email that arrives from my ISP. K9 should absolutely generate 100% accurate to standards mail headers, however to be useful to me it also requires an amount of tolerance of mail headers that fail to meet those stringent standards. virgin media are not a small ISP in some back street, they are one of the UK's biggest isps owned by Comcast. If I'm having this issue you can be certain that many others are also having it and are discarding the app due to its inability to interpret these dates. I understand that the source of the problem is not k9 mail, but the fact is that other mail clients seem to be more tolerant of the issue.

**Comment by MikeGuest** [comment]
And additionally, thanks for those who are putting work into rectifying this. In all likelihood I would resume use of k9 if it could reliably sort my mail. Outlook still can't alert to new emails arriving in anything but the inbox

**Comment by marmistrz** [comment]
TL;DR: we can enforce strictness if we are early adopters and not implemented a widely used protocol

I completely agree that when implementing a new protocol, or a server-to-server protocol one should strictly adhere to the specification to ensure that it's strictly implemented everywhere. But this can be done only in protocol where we can simply disconnect the misbehaving peer. And when there's no implicit industry standard. In case of mail the standard is that most adhere to the standard, but some not.

In case of e-mail it's a completely different matter. That's an old protocol, with everyone having their own quirks. tlen.pl is one of the biggest providers in Poland too. An IMAP client cannot simply disconnect to enforce correctness from the server. Especially, that most e-mail providers largely prefer web interfaces, which can be used to show ads to the user. 

Believe me, that an e-mail provider will not care that a minor MUA doesn't want to adapt to their non-standard quirks. Those who suffer are people who have chosen a quirky mail provider years ago and want to use GPG - there's nothing open source on Android that provides GPG mail.

**Comment by philipwhiuk** [comment]
> virgin media

The ones from @MikeGuest look valid to me - I'll take a look why they aren't parsing. Virgin Media's not owned by Comcast by the way. Either way if they are generating invalid headers then they should be told to fix it.

As for the actual invalid Polish ones, as cketti said, we'll probably implement something, although I'm loathe to just roll over complete - and #1356 has ideas on that

> I completely agree that when implementing a new protocol, or a server-to-server protocol one should strictly adhere to the specification to ensure that it's strictly implemented everywhere.

IMAP did specify it. It's right there in the standard. This isn't about protocol specification it's about power. At minimum we should be flagging that the provider is misbehaving.

**Comment by cketti** [comment]
Sorting by "Date" will use the value of the `Date` header.
Sorting by "Arrival" will use the `INTERNALDATE` value from the IMAP server.

If messages seem out of order (and `Date` headers are valid) it's maybe because you're sorting by 'arrival date'.

**Comment by cketti** [comment]
See #1356
